### ISSUE TITLE
v2.0.0 producer stability observations

### ISSUE THREAD

**Issue by robrigo** [EOSIO/eos#8446]
Our Telos producer node is experiencing intermittent failure.  The producer node is on Ubuntu 18.04 / nodeos v2.0.0 using the pre-build deb packages available under releases. This producer was started using [a v2.0.0 snapshot I generated](https://telos-snapshots.nyc3.cdn.digitaloceanspaces.com/2020-01-14_telos_v2.0.0-snapshot.bin.zip) from an internal v2.0.0 state history node.

I had it set to use eos-vm-jit and eos-vm for a bit, and eventually switched back to wabt but the issue still remains, so I don't think it matters which EOS VM is configured. I'm not sure where the problem lies exactly but I believe it is in the networking code.

The most recent failure started on a freshly booted nodeos with the following errors (not during our round):
```
Jan 15 18:15:50 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:15:50.053 nodeos    producer_plugin.cpp:390       on_incoming_block    ] Received block a131fceddbab6488... #68663114 @ 2020-01-15T18:15:50.000 signed by telosgreenbp [trxs: 0, lib: 68662785, conf: 0, latency: 53 ms]
Jan 15 18:15:50 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:15:50.553 nodeos    producer_plugin.cpp:390       on_incoming_block    ] Received block c993f6cb38a2f8fd... #68663115 @ 2020-01-15T18:15:50.500 signed by telosgreenbp [trxs: 0, lib: 68662785, conf: 0, latency: 53 ms]
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:51.176 nodeos    controller.cpp:1918           apply_block          ] e.to_detail_string(): 3030000 block_validate_exception: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: Block ID does not match
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {"producer_block_id":"0417b74c24fe2b32287f9cc84d82a4ff38e1a1819fde82a0ea723f653120c138","validator_block_id":"0417b74c1732f35c0852ec7228e1c00a5b922d46c09978a9df80387eca5c1338"}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1907 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: warn  2020-01-15T18:15:51.177 nodeos    controller.cpp:1981           push_block           ] 3030000 block_validate_exception: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: Block ID does not match
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {"producer_block_id":"0417b74c24fe2b32287f9cc84d82a4ff38e1a1819fde82a0ea723f653120c138","validator_block_id":"0417b74c1732f35c0852ec7228e1c00a5b922d46c09978a9df80387eca5c1338"}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1907 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1922 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:51.177 nodeos    producer_plugin.cpp:372       on_incoming_block    ] 3030000 block_validate_exception: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: Block ID does not match
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {"producer_block_id":"0417b74c24fe2b32287f9cc84d82a4ff38e1a1819fde82a0ea723f653120c138","validator_block_id":"0417b74c1732f35c0852ec7228e1c00a5b922d46c09978a9df80387eca5c1338"}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1907 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1922 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: rethrow
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1981 push_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:51.177 nodeos    net_plugin.cpp:2946           process_signed_block ] ["95.216.228.50:9876 - c934798" 95.216.228.50:9876]bad signed_block 68663116 24fe2b32287f9cc8...: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:51.177 nodeos    net_plugin.cpp:2947           process_signed_block ] block_validate_exception accept block #68663116 syncing from 95.216.228.50:9876 - c934798
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:15:51.177 net-0     net_plugin.cpp:1059           operator()           ] Sending handshake generation 2 to 95.216.228.50:9876 - c934798, lib 68662785, head 68663115, id c993f6cb38a2f8fd
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:51.210 nodeos    controller.cpp:1918           apply_block          ] e.to_detail_string(): 3030000 block_validate_exception: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: Block ID does not match
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {"producer_block_id":"0417b74c24fe2b32287f9cc84d82a4ff38e1a1819fde82a0ea723f653120c138","validator_block_id":"0417b74c1732f35c0852ec7228e1c00a5b922d46c09978a9df80387eca5c1338"}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1907 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: warn  2020-01-15T18:15:51.210 nodeos    controller.cpp:1981           push_block           ] 3030000 block_validate_exception: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: Block ID does not match
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {"producer_block_id":"0417b74c24fe2b32287f9cc84d82a4ff38e1a1819fde82a0ea723f653120c138","validator_block_id":"0417b74c1732f35c0852ec7228e1c00a5b922d46c09978a9df80387eca5c1338"}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1907 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1922 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:51.210 nodeos    producer_plugin.cpp:372       on_incoming_block    ] 3030000 block_validate_exception: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: Block ID does not match
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {"producer_block_id":"0417b74c24fe2b32287f9cc84d82a4ff38e1a1819fde82a0ea723f653120c138","validator_block_id":"0417b74c1732f35c0852ec7228e1c00a5b922d46c09978a9df80387eca5c1338"}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1907 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1922 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: rethrow
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1981 push_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:51.211 nodeos    net_plugin.cpp:2946           process_signed_block ] ["p2p.telos.dutcheos.io:9876 - 65f27bb" 190.2.155.143:9876]bad signed_block 68663116 24fe2b32287f9cc8...: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:51.211 nodeos    net_plugin.cpp:2947           process_signed_block ] block_validate_exception accept block #68663116 syncing from p2p.telos.dutcheos.io:9876 - 65f27bb
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:15:51.211 net-0     net_plugin.cpp:1059           operator()           ] Sending handshake generation 3 to p2p.telos.dutcheos.io:9876 - 65f27bb, lib 68662785, head 68663115, id c993f6cb38a2f8fd
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:51.243 nodeos    controller.cpp:1918           apply_block          ] e.to_detail_string(): 3030000 block_validate_exception: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: Block ID does not match
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {"producer_block_id":"0417b74c24fe2b32287f9cc84d82a4ff38e1a1819fde82a0ea723f653120c138","validator_block_id":"0417b74c1732f35c0852ec7228e1c00a5b922d46c09978a9df80387eca5c1338"}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1907 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: warn  2020-01-15T18:15:51.243 nodeos    controller.cpp:1981           push_block           ] 3030000 block_validate_exception: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: Block ID does not match
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {"producer_block_id":"0417b74c24fe2b32287f9cc84d82a4ff38e1a1819fde82a0ea723f653120c138","validator_block_id":"0417b74c1732f35c0852ec7228e1c00a5b922d46c09978a9df80387eca5c1338"}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1907 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1922 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:51.243 nodeos    producer_plugin.cpp:372       on_incoming_block    ] 3030000 block_validate_exception: Block exception
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: Block ID does not match
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {"producer_block_id":"0417b74c24fe2b32287f9cc84d82a4ff38e1a1819fde82a0ea723f653120c138","validator_block_id":"0417b74c1732f35c0852ec7228e1c00a5b922d46c09978a9df80387eca5c1338"}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1907 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1922 apply_block
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]: rethrow
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     {}
Jan 15 18:15:51 ip-10-0-0-184 start.sh[23735]:     nodeos  controller.cpp:1981 push_block
```

After this I saw a series of bad_signed_block and unlinkable block issues:
```
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:52.188 nodeos    net_plugin.cpp:2946           process_signed_block ] ["ip-10-0-0-197:1337 - 498d4ba" 10.0.0.197:1337]bad signed_block 68663116 24fe2b32287f9cc8...: Block exception
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:52.188 nodeos    net_plugin.cpp:2947           process_signed_block ] block_validate_exception accept block #68663116 syncing from ip-10-0-0-197:1337 - 498d4ba
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:52.188 nodeos    net_plugin.cpp:2943           process_signed_block ] ["ip-10-0-0-197:1337 - 498d4ba" 10.0.0.197:1337]bad signed_block 68663117 1b094afd4ece6017...: Unlinkable block
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:52.188 nodeos    net_plugin.cpp:2943           process_signed_block ] ["ip-10-0-0-197:1337 - 498d4ba" 10.0.0.197:1337]bad signed_block 68663118 090fb6e500dfec74...: Unlinkable block
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:52.348 nodeos    net_plugin.cpp:2946           process_signed_block ] ["chia:9876 - 2aba8a9" 206.124.132.247:9876]bad signed_block 68663116 24fe2b32287f9cc8...: Block exception
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:52.348 nodeos    net_plugin.cpp:2947           process_signed_block ] block_validate_exception accept block #68663116 syncing from chia:9876 - 2aba8a9
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:52.348 nodeos    net_plugin.cpp:2943           process_signed_block ] ["chia:9876 - 2aba8a9" 206.124.132.247:9876]bad signed_block 68663117 1b094afd4ece6017...: Unlinkable block
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: error 2020-01-15T18:15:52.348 nodeos    net_plugin.cpp:2943           process_signed_block ] ["chia:9876 - 2aba8a9" 206.124.132.247:9876]bad signed_block 68663118 090fb6e500dfec74...: Unlinkable block
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:15:52.349 net-1     net_plugin.cpp:1059           operator()           ] Sending handshake generation 7 to chia:9876 - 2aba8a9, lib 68662785, head 68663115, id c993f6cb38a2f8fd
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:15:52.350 net-1     net_plugin.cpp:1059           operator()           ] Sending handshake generation 8 to chia:9876 - 2aba8a9, lib 68662785, head 68663115, id c993f6cb38a2f8fd
Jan 15 18:15:52 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:15:52.350 net-1     net_plugin.cpp:1059           operator()           ] Sending handshake generation 9 to chia:9876 - 2aba8a9, lib 68662785, head 68663115, id c993f6cb38a2f8fd
```

Followed by a short round of producing blocks from eosiodetroit due to what I believe is "old state head catchup becoming in sync" blocking production:
```
Jan 15 18:17:09 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:17:09.577 nodeos    producer_plugin.cpp:390       on_incoming_block    ] Received block a83756220488a931... #68663265 @ 2020-01-15T18:17:05.500 signed by eosdublinwow [trxs: 0, lib: 68662929, conf: 0, latency: 4077 ms]
Jan 15 18:17:09 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:17:09.587 net-0     net_plugin.cpp:1364           set_state            ] old state head catchup becoming in sync
Jan 15 18:17:09 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:17:09.588 net-0     net_plugin.cpp:1059           operator()           ] Sending handshake generation 3 to 95.216.228.50:9876 - c934798, lib 68662929, head 68663265, id a83756220488a931
Jan 15 18:17:10 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:17:10.001 nodeos    producer_plugin.cpp:1965      produce_block        ] Produced block ac70602450399fdb... #68663266 @ 2020-01-15T18:17:10.000 signed by eosiodetroit [trxs: 7, lib: 68662941, confirmed: 240]
Jan 15 18:17:10 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:17:10.500 nodeos    producer_plugin.cpp:1965      produce_block        ] Produced block f2b31e8b550262f4... #68663267 @ 2020-01-15T18:17:10.500 signed by eosiodetroit [trxs: 1, lib: 68662941, confirmed: 0]
Jan 15 18:17:11 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:17:11.000 nodeos    producer_plugin.cpp:1965      produce_block        ] Produced block 59d893b4e38aa728... #68663268 @ 2020-01-15T18:17:11.000 signed by eosiodetroit [trxs: 0, lib: 68662941, confirmed: 0]
Jan 15 18:17:11 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:17:11.500 nodeos    producer_plugin.cpp:1965      produce_block        ] Produced block 46f58f777c06f952... #68663269 @ 2020-01-15T18:17:11.500 signed by eosiodetroit [trxs: 1, lib: 68662941, confirmed: 0]
Jan 15 18:17:12 ip-10-0-0-184 start.sh[23735]: info  2020-01-15T18:17:12.177 nodeos    producer_plugin.cpp:390       on_incoming_block    ] Received block e82392fb0ed76936... #68663270 @ 2020-01-15T18:17:12.000 signed by eosphereiobp [trxs: 0, lib: 68662953, conf: 232, latency: 177 ms]
```

Afterwards, the node recovered for a bit and started producing full rounds again. 

My internal network peers are running eos-vm-jit with oc-threads enabled, and I set the net threads to 2 for all internal peers as well.

I've attached a snapshot of the logs that include all of the details:

[telos-v2.0.0-producer.log](https://github.com/EOSIO/eos/files/4066742/telos-v2.0.0-producer.log)

**Comment by robrigo** [comment]
Shortly after posting this I noticed these issues returning, so it seems to come and go.

**Comment by robrigo** [comment]
Also seeing this error pop up randomly: `FC Exception encountered while processing chain.send_transaction`

**Comment by heifner** [comment]
Are you only seeing this on your nodes running the `state_history_plugin` ?

**Comment by robrigo** [comment]
Hi @heifner,

This node doesn't have state history turned on. It has the following plugins turned on:
```
plugin = eosio::producer_plugin
plugin = eosio::chain_plugin
plugin = eosio::chain_api_plugin
plugin = eosio::net_plugin
plugin = eosio::http_plugin
```

**Comment by heifner** [comment]
Did you run any 2.0 rc releases before this? If so did you see this there?

**Comment by robrigo** [comment]
I have run all of the rc's in testnet and have been running the RCs on our prod state history node which also generated the snapshot being used on this node.

**Comment by robrigo** [comment]
I replayed the block.log from the ship node on the producer node described in the original post, instead of starting it by a snapshot. I noticed that I was still encountering the unlinkable blocks issue, but removing a specific public p2p-peer-address seems to have made it go away.

**Comment by robrigo** [comment]
I have these external peers spread out across our internal mesh, along with my internal peers. they each have a sample of this list:

[telos-peers.txt](https://github.com/EOSIO/eos/files/4077681/telos-peers.txt)

**Comment by villesundell** [comment]
I am also experiencing this same issue. My API nodes work well with 2.0.0, but backup BP does not (main BP is still 1.8x). I haven't noticed any difference regarding peers: I have run my recent tests with one peer only (some tests with my own 2.0.0 seed node, some tests with GoodBlock's 1.8.x node), and still experiencing this issue.

However, for some reason enabling "chain_api" plugin helps!

So, if there are only "chain" and "producer" plugins enabled, I experience this issue. But when I enable "chain_api" plugin, everything seems to work.

**Comment by villesundell** [comment]
Since updating to 2.0.1 I haven't experienced this issue :+1:

**Comment by heifner** [comment]
Closing as this seems to have been fixed in later 2.0.x versions.
